File: ffi/ruby/src/ffi.js
/**
 * Created by dp on 2017/5/24.
 */

var ffi = require('ffi');

// process.platform
// os.platform()

var lib
if(process.platform == 'win32'){
    lib = ffi.Library('target\\release\\ruby.dll', {
        'process': ['void', []]
    })
}else if(process.platform == 'linux'){
    lib = ffi.Library('target/release/libruby.so', {
        'process': ['void', []]
    })
}else{
    lib = ffi.Library('target/release/libruby.so', {
        'process': ['void', []]
    })
}

lib.process()

console.log('done!')

File: ffi/src/ffi.rb
require 'ffi'

def os_family
  case RUBY_PLATFORM
    when /linux/
      :linux
    when /ix/i, /ux/i, /gnu/i,
        /sysv/i, /solaris/i,
        /sunos/i, /bsd/i
      :unix
     when /darwin|mac os/
      :macosx
    when /win/i, /ming/i
      :windows
    else
      :other
  end
end

Dir.chdir('..')

module Hello
  extend FFI::Library
  ffi_lib os_family == :windows ? 'target/release/ruby_process.dll' : os_family == :linux || os_family == :unix ?
      'target/release/libruby_process.so' : os_family == :macosx ?
      'target/release/libruby_process.dylib' : 'target/release/libruby_process.so'
  attach_function :process, [], :void
end

Hello.process

puts 'done!'

puts os_family
